### PR TITLE
R packaging improvements

### PR DESCRIPTION
**Pull request by richardbeare**
This is a series of updates focused on getting the R package tests working. There are basically 2 parts - documentation generation and namespace/swig modifications.

The documentation changes are partly in the generation script - eliminating empty fields and moving some sections around to satisfy the rules. There are also some manually generated manual files to satisfy package tests while avoiding clutter of the indexes.

There are now some additional steps in generating the R code. Unfortunately some of the R code generated by swig is wrong. We hadn't noticed because that code isn't used. However the package checks uncover the issues. Most of the errors are to do with delete methods, and they are removed via an R grep. Another error is fixed by overwriting an incorrect function manually. The namespace processing script has also been extended to allow a manually selected list of unused functions to be excluded from the export list.

Finally, automated processing us used to create a documentation file for the '$' method. This is to avoid cluttering the index while keeping the checker happy.

### CODE REVIEW

**Comment by richardbeare** [comment]
Rebased - awaiting testing.

**Comment by richardbeare** [comment]
Note - the prebuilt documentation will need to be refreshed after this update is merged.

File: Utilities/GenerateDocs/README [hunk 1]
```diff
@@ -15,9 +15,17 @@ the SwigDocUpdate.sh script can be run to convert the SimpleITK
 Doxygen XML into Swig .i files so that when SimpleITK is built, inline
 documenation for some wrapped languages will be available.
 
+Building the documentation target from the SuperBuild framework requires:
+        The correct cmake flags: -DBUILD_DOXYGEN:BOOL=ON -DBUILD_DOCUMENTS:BOOL=ON
```
Comment (blowekamp) [inline Utilities/GenerateDocs/README]: The "-DBUILD_DOCUMENTS:BOOL=ON"flag  is not needed and should be removed.

There is also a super build target "doc" could be run.

File: Utilities/GenerateDocs/README [hunk 2]
```diff
@@ -15,9 +15,17 @@ the SwigDocUpdate.sh script can be run to convert the SimpleITK
...
+        Run "make Documentation" in the SimpleITK-build directory.
+        This will create the SimpleITK-build/Documentation/xml content.
+
 Important end use scripts (see script files for more details:
         JSONDocUpdate.sh - Copies ITK Doxygen into SimpleITK JSON
-        SwigDocUpdate.sh - Copies SimpleITK Doxygen into Swig .i files
+        SwigDocUpdate.sh - Copies SimpleITK Doxygen into Swig .i files and
+                           R .Rd files.
+
+        SwigDocUpdate requires python, with the regex library.
```
Comment (blowekamp) [inline Utilities/GenerateDocs/README]: Why is the regex library needed over the standard "[re](https://docs.python.org/2.7/library/re.html)" Python module.

File: Utilities/GenerateDocs/doxy2swig.py [hunk 3]
```diff
@@ -451,8 +451,36 @@ def __init__(self, src, javaFlag=0):
...
+        self.mathstuff3 = re.compile(r"\\end\{array\}")
+        # a complex recursive regular expression, to deal with formula
+        # inside mbox and text structures
+        self.mathstuff4 = regex.compile(r"\\mbox({((?>[^}{]*(?1)?)*)})", flags=regex.V1)
+        self.mathstuff5 = regex.compile(r"\\text({((?>[^}{]*(?1)?)*)})", flags=regex.V1)
+        # the special doxygen tags - note - not greedy
+        self.mathstuff6 = re.compile(r"\\f\$(.+?)\\f\$")
+        # alignment tags
+        self.mathstuff7 = re.compile(r" & ")
+    def filterLatexMath(self, txt):
```
Comment (blowekamp) [inline Utilities/GenerateDocs/doxy2swig.py]: PEP8: space before function definition. and 

"""
Comment about what the function is doing.
"""

File: Utilities/GenerateDocs/doxy2swig.py [hunk 4]
```diff
@@ -725,36 +756,51 @@ def do_memberdef(self, node):
                     self.add_text('\\name{%s}\n' %(name))
                     self.add_text('\\alias{%s}\n' %(name))
                     self.add_text('\\description{\n')
-                    self.generic_parse(others['briefdescription'])
+                    if others['briefdescription'] is None or \
+                       others['briefdescription'].firstChild.data.isspace() or \
+                       others['briefdescription'].firstChild.data == '':
+                       self.add_text('See class definition - most likely nameImageFilter')
```
Comment (blowekamp) [inline Utilities/GenerateDocs/doxy2swig.py]: Did you intend to use the "name" variable in the text output?

File: Wrapping/R/CMakeLists.txt [hunk 5]
```diff
@@ -125,8 +125,9 @@ endif()
 add_custom_command( TARGET ${SWIG_MODULE_SimpleITK_TARGET_NAME}
   POST_BUILD
   COMMAND ${CMAKE_COMMAND} -E make_directory ${CMAKE_CURRENT_BINARY_DIR}/R_libs
-  COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_BINARY_DIR}/SimpleITK.R ${CMAKE_CURRENT_BINARY_DIR}/Packaging/SimpleITK/R/
+  #COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_BINARY_DIR}/SimpleITK.R ${CMAKE_CURRENT_BINARY_DIR}/Packaging/SimpleITK/R/
```
Comment (blowekamp) [inline Wrapping/R/CMakeLists.txt]: delete the commented out command

File: Wrapping/R/processSwigCode.R [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+processSwig <- function(swigsrc, swigdest, docdest) {
+  code <- readLines(swigsrc)
+  code <- grep("^setMethod\\('delete',", code, value=TRUE, invert=TRUE)
+  writeLines(code, swigdest)
+  dollarmethods <- grep("^setMethod\\('\\$',", code, value=TRUE)
+  writeAlias(dollarmethods, docdest)
+}
+args <- commandArgs( TRUE )
```
Comment (blowekamp) [inline Wrapping/R/processSwigCode.R]: Do we have documentation on the expected arguments or how to use this script?

File: Utilities/GenerateDocs/README (hunk 2 above)
Comment (richardbeare) [reply Utilities/GenerateDocs/README]: Evil recursive regular expression for matching bracket pairs.

**Comment by richardbeare** [comment]
Amended commit in response to comments

**Comment by blowekamp** [comment]
Great! Thank you for making those changes. This ready to be merged?

**Comment by richardbeare** [comment]
Yep, good to go.

Are you able to deal with the generated doc files?

**Comment by blowekamp** [comment]
What do yo mean am I able to deal with the generated docs?

**Comment by richardbeare** [comment]
The manual pages are downloaded from somewhere as part SimpleITKRpackageData, so the online archive needs to be updated with the new versions.

**Comment by blowekamp** [comment]
topic merged into next

**Pull request merged**